internal/testrunner/runners/system/servicedeployer/custom_agent.go

```diff
@@ -0,0 +1,131 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package servicedeployer
+
+import (
+	_ "embed"
+	"os"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	"github.com/elastic/elastic-package/internal/install"
+	"github.com/elastic/elastic-package/internal/logger"
+)
+
+//go:embed custom-agent-base-config.yml
+var customAgentYml string
+
+// CustomAgentDeployer knows how to deploy a custom elastic-agent defined via
+// a Docker Compose file.
+type CustomAgentDeployer struct {
+	cfg string
+	cd  *DockerComposeServiceDeployer
+}
+
+type deployedCustomAgent struct {
+	cfg string
+	*dockerComposeDeployedService
+}
+
+var _ DeployedService = &deployedCustomAgent{}
+
+// NewCustomAgentDeployer returns a new instance of a deployedCustomAgent.
+func NewCustomAgentDeployer(cfgPath string, ymlPaths []string, sv ServiceVariant) (*CustomAgentDeployer, error) {
+	path, err := createCustomAgentYaml(cfgPath)
+	if err != nil {
+		return nil, err
+	}
+
+	appConfig, err := install.Configuration()
+	if err != nil {
+		return nil, errors.Wrap(err, "can't read application configuration")
+	}
+
+	cd, err := newDockerComposeServiceDeployer(
+		append(ymlPaths, path),
+		appConfig.StackImageRefs(install.DefaultStackVersion).AsEnv(),
+		sv,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CustomAgentDeployer{
+		cd:  cd,
+		cfg: path,
+	}, nil
+}
+
+// SetUp sets up the service and returns any relevant information.
+func (d *CustomAgentDeployer) SetUp(inCtxt ServiceContext) (DeployedService, error) {
+	ds, err := d.cd.SetUp(inCtxt)
+	if err != nil {
+		return nil, err
+	}
+
+	dds, ok := ds.(*dockerComposeDeployedService)
```

Comment:
Hm... I used to go without casting as [here](https://github.com/elastic/elastic-package/blob/main/internal/testrunner/runners/system/servicedeployer/kubernetes.go#L103). Is it mandatory really here?

Reply:
To have access to the internal context to override it we need the specific type, since the interface does not expose it.

Reply:
I'm not sure if it doesn't loosen the dependency between Docker service deployer and Custom Agent deployer. Take a look at the [terraform](https://github.com/elastic/elastic-package/blob/main/internal/testrunner/runners/system/servicedeployer/terraform.go#L32) deployer. It looks like you need to define the Compose project. WDYT?